### ISSUE TITLE
n-gram combinations

### ISSUE THREAD

**Issue by paceaux** [paceaux/methodius#15]
Need ability to discover n-grams that commonly occur with other n-grams. 

i.e. How would I discover "tion"

e.g.
"ion"
"tio"
nation => nat ati tio ion
vacation => vac aca cat tio ion
station => sta ati tio ion

**Comment by paceaux** [comment]
Note:

discovering tion is a case of knowing siblings

after-sibling for `tio` is `ion`
before-sibling for `ion` is `tio`

So they're vectors?

loop through `nation`
`[na, at, ti, io, on]`

```JavaScript
io ={
 after: 'on'
 before: 'ti'
}

on = {
after: ''
before: 'io'
}
```

**Comment by paceaux** [comment]
note:

discovering this automatically vs manually are diff tasks

```JavaScript
      const words = ['revolution', 'nation'];
      const ngramCollections = Methodius.getNgramCollections(words, 2);
      const onSiblings = Methodius.getNgramSiblings('io', ngramCollections);
```
this produces 

Map(
'before': Map(
  'ti': 2,
),
'after': Map(
   'on': 2,
  ),
)

```

This means that by searching for 'io' we can discover that 'ti' comes before, and 'io' comes after.

**Comment by paceaux** [comment]
Different approaches for this process

### A UnionReducer
We already have "comparison functions":
* union (takes two arrays and merges them into one where shared items only appear once)
* disjunctive union (takes two arrays and returns only items unique to each array)
* intersection (takes two arrays and returns only the items shared amongst both)

A UnionReducer would be a function that allows more than two arguments and recursively uses `getUnion` to join all of those items into one thing. 

This would merge `[ti, io, on]` into `tion`. 

The idea would be that we'd merge NGrams into larger parts and then search to see if those larger parts exist. 

the problem is that this could be a very expensive operation if we're dealing with lots of text. So we'd have to make some sort of intelligent decision to only do this on ngrams of a given frequency. 

### Vectorization
1. use `getNGramSiblings` to generate a map of maps which contains _every_ Ngram's sibling. 
2. Then loop through the entire collection of siblings and find ones where their `after` matches another's `before`. 

It amounts to discovering that two items are related to each other. 

This, again, would be an expensive operation. First b/c we have to generate the siblings, second because we then have to loop through it to make determinations. 

### An NGram Tree
We recursively ngramify every word, forming a tree:

```JavaScript
revolution =>
{
'revol' : {
  'rev' : {
     're',
      'ev',
   },
  'vol': {
     'vo',
     'ol'
  }
 },
'ution' : {
  'uti' : {
    'ut',
    'ti'
   },
  'ion': {
    'io', 
    'on'
   }
}
}
```

yes, it would be hard to do this, and there's edge cases involved. But by nature of building a tree we would actually be able to work top-down instead of bottom up.

**Comment by paceaux** [comment]
This was a massive amount of work and it shipped with two big features. 

1. there's an ngram tree as described above; it's possible to break down a word into a series of recursively smaller ngrams. It's also possible to search that tree. 
2. There's a "get related ngrams" that is actually quite simple and dumb. Feed it a list of ngrams and a list of words. If there's ngram1 and ngram2 occuring together, it gets added to a list. In the `Methodius` instance class, this uses only top ngrams. But ... it could be any. 

released as part of 2.0
